onadata/libs/utils/project_utils.py

```diff
@@ -138,3 +153,177 @@ def _set_project_perms():
+    resp = session.post(asset_url, json=payload)
+    if resp.status_code != 200:
+        raise ExternalServiceRequestError(
+            f"Failed to assign permissions for {asset_id}: {resp.status_code}"
+        )
+    return resp
+
+
+@app.task(bind=True, max_retries=3)
+def propagate_project_permissions_async(
+    self,
+    project_id: int,
+    headers: Optional[dict] = None,
+    use_asset_owner_auth: bool = True,
+):
+    """
+    Asynchronously propagates Project Permissions to the Formbuilder assets
+    within the project
+    """
+    try:
+        project = Project.objects.get(id=project_id)
+        propagate_project_permissions(project, headers, use_asset_owner_auth)
+    except (Project.DoesNotExist, ExternalServiceRequestError) as e:
+        if self.request.retries > 3:
+            msg = f"Failed to propagate asset permissions for Project {project_id}"
+            report_exception(msg, e, sys.exc_info())
+        self.retry(exc=e, countdown=60 * self.request.retries)
+
+
+def propagate_project_permissions(
+    project: Project, headers: Optional[dict] = None, use_asset_owner_auth: bool = True
+) -> None:
+    """
+    Propagates Project permissions to external services(KPI)
+    """
+    if getattr(settings, "KPI_INTERNAL_SERVICE_URL", None):
+        service_url = settings.KPI_INTERNAL_SERVICE_URL
+
+        # Create request session for the Internal KPI calls
+        session = requests.session()
+        session.mount(
+            "http",
+            HTTPAdapter(
+                max_retries=Retry(
+                    total=5,
+                    backoff_factor=2,
+                    method_whitelist=["GET", "POST", "DELETE"],
+                    status_forcelist=[502, 503, 504],
+                )
+            ),
+        )
+
+        if headers:
+            session.headers.update(headers)
+
+        # Retrieve users who have access to the project
+        collaborators: List[str] = [
+            username
+            for username, data in get_project_users(project).items()
+            if data.get("role") != "owner" and data.get("role") in ["editor"]
+        ]
+
+        if is_organization(project.organization.profile):
+            owners_team = Team.objects.get(
+                name=f"{project.organization.username}#{Team.OWNER_TEAM_NAME}"
+            )
+            owners_team = list(
+                owners_team.user_set.filter(
+                    ~Q(username=project.organization.username)
+                ).values_list("username", flat=True)
+            )
+            collaborators += owners_team
+            collaborators = list(set(collaborators))
+
+        form_ids: List[int] = list(
+            project.xform_set.filter(deleted_at__isnull=True).values_list(
+                "id", flat=True
+            )
+        )
+
+        # Propagate permissions for XForms that were published by
+        # Formbuilder
+        for form_id in form_ids:
+            asset: XForm = XForm.objects.get(id=form_id)
```

Comment:
We should add a try catch block here so that if the xform doesnt exist in replica it can be fetched from master

Reply:
I think that might not be something we'd need to handle since [this](https://github.com/onaio/onadata/pull/2336/files/e246be0987dd007094b7e3b606b11591d912d7f6#diff-b788ff4632a7e33981fda5c4e9b7dfe6f2b550fcf2d60b65126fe4d808cb0030R275) line should only return XForm objects that are available in the moment